File: logic/src/main/java/a75f/io/logic/pubnub/AlertDefinitionHandler.java
package a75f.io.logic.pubnub;

import com.fasterxml.jackson.annotation.JsonAutoDetect;
import com.fasterxml.jackson.annotation.PropertyAccessor;
import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.google.gson.JsonObject;

import org.json.JSONObject;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Iterator;

import a75f.io.alerts.AlertDefinition;
import a75f.io.alerts.AlertManager;
import a75f.io.alerts.HttpUtil;
import a75f.io.logger.CcuLog;
import a75f.io.logic.Globals;
import a75f.io.logic.L;

public class AlertDefinitionHandler
{
    public static final String CMD = "alertDefinition";
    
    public static void handleMessage(JsonObject msgObject)
    {
        String alertGUID = msgObject.get("alert_def_id").getAsString();
        ArrayList<AlertDefinition> alertList = null;
        try
        {
            JSONObject postData = new JSONObject();
            postData.put("_id", alertGUID);
    
            String alertDef = HttpUtil.sendRequest("readDef", postData.toString());
            CcuLog.d(L.TAG_CCU_PUBNUB," alertDef "+alertDef);
            ObjectMapper objectMapper = new ObjectMapper();
            objectMapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
            objectMapper.setVisibility(PropertyAccessor.FIELD, JsonAutoDetect.Visibility.ANY);
            AlertDefinition[] pojos = objectMapper.readValue(alertDef, AlertDefinition[].class);
            alertList = new ArrayList<>(Arrays.asList(pojos));
    
            Iterator iterator = alertList.iterator();
            while(iterator.hasNext())
            {
                AlertDefinition a = (AlertDefinition) iterator.next();
                if (a.validate() == false) {
                    CcuLog.d(L.TAG_CCU_PUBNUB, " Invalid Alert Definition "+a.toString());
                    iterator.remove();
                }
            }
            if (alertList.size() > 0)
            {
                AlertManager.getInstance(Globals.getInstance().getApplicationContext()).addAlertDefinition(alertList);
            }
        }
        catch (Exception e)
        {
            CcuLog.d(L.TAG_CCU_PUBNUB,"alertDef Parse Failed "+e.getMessage());
            e.printStackTrace();
        }
        
    }
}

File: logic/src/main/java/a75f/io/logic/pubnub/AlertDefinitionHandler.kt
@file:JvmName("AlertDefinitionHandler")

package a75f.io.logic.pubnub

import a75f.io.alerts.AlertManager
import a75f.io.alerts.cloud.ServiceGenerator
import a75f.io.api.haystack.CCUHsApi
import a75f.io.logger.CcuLog
import a75f.io.logic.L
import com.google.gson.JsonObject
import io.reactivex.rxjava3.schedulers.Schedulers

const val CMD = "alertDefinition"

/**
 * Handles the alert definition pubnub msg, which has one function, to
 * get the alert definition from the service and save it if it is valid.
 *
 * Follows the exact logic of previous AlertDefinitionHandler, except:
 *  - uses AlertService (v2) instead of HttpUtil (v1)
 *  - needs siteId for v2
 *  - gets a single alertDef instead of a List
 *  - logs more errors.
 *
 * @author tcase@75f.io
 * Created on 3/10/21.
 */
fun handleMessage(msgObject: JsonObject) {

   val alertGUID = msgObject["alert_def_id"].asString
   val siteId = CCUHsApi.getInstance().globalSiteId?.removePrefix("@")
                  ?: return CcuLog.w(L.TAG_CCU_PUBNUB, "No siteId in handle alert def message")

   // Alert service will have been initialized at startup
   val alertsService = ServiceGenerator.instance.alertsService

   // Note: Following existing logic, we are not saving the ref to the disposable here.  We should, if
   // we were to follow best practice to a T, keep the disposable and dispose it if the app needs to exit
   // while this logic is in flight.
   alertsService.getCustomDefinition(siteId, alertGUID)
      .subscribeOn(Schedulers.io())
      .subscribe(
         { alertDef ->
            if (alertDef.validate()) {
               AlertManager.getInstance().addAlertDefinition(alertDef)
            } else {
               CcuLog.d(L.TAG_CCU_PUBNUB, "Invalid Alert Definition $alertDef")
            }
         },
         { error -> CcuLog.w(L.TAG_CCU_PUBNUB, "Error fetching alert def from server", error) }
      )
}
